File: 20190519python/0092. Reverse Linked List II.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reverseBetween(self, head: ListNode, m: int, n: int) -> ListNode:
        node=ListNode(0)
        node.next=head
        p=node
        while m>1:
            p=p.next
            m-=1
            n-=1
        
        p.next=self.reverseToN(p.next,n)
        return node.next
    
    def reverseToN(self,head,n):
        cur=head
        while n>1:
            cur=cur.next
            n-=1
        
        prev=cur.next
        cur.next=None
        while head:
            cur=head
            head=head.next
            cur.next=prev
            prev=cur
        return prev
        

File: archives/C#/0092. Reverse Linked List II.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode ReverseBetween(ListNode head, int m, int n) {
        ListNode node=new ListNode(0);
        node.next=head;
        ListNode p=node;
        while(m-1>0){
            p=p.next;
            m--;
            n--;
        }
        p.next=ReverseFirstN(p.next,n);
        return node.next;
    }
    public ListNode ReverseFirstN(ListNode head,int n){
        ListNode p=head;
        int m=n;
        while(m>1){
            p=p.next;
            m--;
        }
        
        ListNode prev=p.next;
        p.next=null;
        ListNode cur;
        
        while(n>0){
            n--;
            cur=head;
            head=head.next;
            cur.next=prev;
            prev=cur;
        }
        return prev;
    }
}
